executor/windowfuncs/window_funcs.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package windowfuncs
+
+import (
+	"errors"
+
+	"github.com/pingcap/tidb/executor/aggfuncs"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// WindowFunc is the interface for processing window functions.
+type WindowFunc interface {
+	// ProcessOneChunk processes one chunk.
+	ProcessOneChunk(sctx sessionctx.Context, rows []chunk.Row, dest *chunk.Chunk) ([]chunk.Row, error)
+	// ExhaustResult exhausts result to the result chunk.
+	ExhaustResult(sctx sessionctx.Context, rows []chunk.Row, dest *chunk.Chunk) ([]chunk.Row, error)
+	// HasRemainingResults checks if there are some remained results to be exhausted.
+	HasRemainingResults() bool
+}
+
+// aggWithoutFrame deals with agg functions with no frame specification.
+type aggWithoutFrame struct {
+	result   aggfuncs.PartialResult
+	agg      aggfuncs.AggFunc
+	remained int64
+}
+
+// ProcessOneChunk implements the WindowFunc interface.
+func (wf *aggWithoutFrame) ProcessOneChunk(sctx sessionctx.Context, rows []chunk.Row, dest *chunk.Chunk) ([]chunk.Row, error) {
```

Comment:
we don't need `dest`?

Reply:
In here we do not need an `dest`, but for functions that can be streamed like `row_number` we need it because we can directly write the result in the function.